one tutorial to rule them all

joelgrus opened the pull request:
I started with the PyTorch sequence tagging tutorial: https://pytorch.org/tutorials/beginner/nlp/sequence_models_tutorial.html

My goal was to make a tutorial that should be "the only tutorial you need" for like 95% of our users. You tell me if this accomplishes that.

joelgrus commented on tutorials/tagger/pos_allennlp_bare.py:
```diff
@@ -0,0 +1,130 @@
+from allennlp.modules.text_field_embedders import TextFieldEmbedder, BasicTextFieldEmbedder
+from allennlp.modules.token_embedders import Embedding
+from allennlp.modules.seq2seq_encoders import Seq2SeqEncoder, PytorchSeq2SeqWrapper
+from allennlp.nn.util import get_text_field_mask, sequence_cross_entropy_with_logits
```
that's a lot of imports!

matt-gardner replied:
You get around this in the plain pytorch version by importing `nn` and doing `nn.X`.  You could do the same here, if you wanted to.

schmmd commented:
Closing in favor of https://github.com/allenai/allennlp-website/pull/73.

joelgrus commented:
I need to still merge this into the main tutorials codebase (along with the code required to transform it into the right format)

matt-gardner commented on tutorials/tagger/basic_allennlp_annotated.py:
```diff
@@ -0,0 +1,204 @@
+
+#### As in the original PyTorch tutorial, we'd like to look at the predictions our model generates. AllenNLP contains a <code>Predictor</code> abstraction that takes inputs, converts them to instances, feeds them through your model, and returns JSON-serializable results. Often you'd need to implement your own Predictor, but AllenNLP already has a <code>SentenceTaggerPredictor</code> that works pe ...
+predictor = SentenceTaggerPredictor(model, dataset_reader=reader)
+#### It has a <code>predict</code> method that just needs a sentence and returns (a JSON-serializable version of) the output dict from forward.  Here "tag_scores" will be a (5, 3) array of logits, corresponding to the 3 possible tag for each of the 5 words.
```
3 possible tags

matt-gardner replied:
And when I first read this I was confused at the mismatch between "tag_logits" and "tag_scores".  Also, "tag_scores" in the annotation should probably be in a <code> block.

matt-gardner commented on tutorials/tagger/basic_allennlp_annotated.py:
```diff
@@ -0,0 +1,204 @@
+#### Now that we've implemented a <code>DatasetReader</code> and <code>Model</code>, we're ready to train. We first need an instance of our dataset reader.
+reader = PosDatasetReader()
+#### Which we can use to read in the training data and validation data.
+train_dataset = reader.read('tutorials/tagger/training.txt')
```
And if you make it so that you can just copy and paste this whole thing and run it, I might even mention that up at the top.

joelgrus commented:
ok, I think I addressed all of your comments, except that I didn't change the validation metric because on a dataset this small the training spends a long time with accuracy not improving and so I would have had to set the patience to a really huge number to get it to train

matt-gardner commented on tutorials/tagger/basic_allennlp.py:
```diff
@@ -0,0 +1,202 @@
+    #### Next we need to implement <code>forward</code>, which is where the actual computation happens. Each <code>Instance</code> in your dataset will get (batched with other instances and) fed into <code>forward</code>. The <code>forward</code> method expects dicts of tensors as input, and it expects their names to be the names of the fields in your <code>Instance</code>. In this case we have a ...
+    def forward(self,
+                sentence: Dict[str, torch.Tensor],
+                labels: Dict[str, torch.Tensor] = None) -> torch.Tensor:
```
This type annotation should still be a `Tensor.

joelgrus replied:
fixed

matt-gardner commented on tutorials/tagger/basic_allennlp.py:
```diff
@@ -0,0 +1,202 @@
+
+    #### The other piece we have to implement is <code>_read</code>, which takes a filename and produces a stream of <code>Instance</code>s. Most of the work has already been done in <code>text_to_instance</code>.
+    def _read(self, file_path: str) -> Iterator[Instance]:
+        with open(file_path) as f:
```
Did you actually try copying and pasting this code and running it?  I think you need a `cached_path` here.

joelgrus replied:
I didn't run it with the URLs because the files aren't on GitHub until I merge this PR 😢

joelgrus replied:
fixed
